packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/components/NodeChildrenProvider.java

```diff
@@ -0,0 +1,149 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Obeo.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.sirius.components.diagrams.components;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.eclipse.sirius.components.diagrams.Node;
+import org.eclipse.sirius.components.diagrams.ViewModifier;
+import org.eclipse.sirius.components.diagrams.description.InsideLabelDescription;
+import org.eclipse.sirius.components.diagrams.description.NodeDescription;
+import org.eclipse.sirius.components.diagrams.description.OutsideLabelDescription;
+import org.eclipse.sirius.components.diagrams.renderer.DiagramRenderingCache;
+import org.eclipse.sirius.components.representations.Element;
+import org.eclipse.sirius.components.representations.VariableManager;
+
+/**
+ * Provide the children element of node.
+ *
+ * @author frouene
+ */
+public class NodeChildrenProvider {
+
+    private final NodeComponentProps props;
+
+    public NodeChildrenProvider(NodeComponentProps props) {
+        this.props = Objects.requireNonNull(props);
+    }
+
+    public List<Element> getNodeChildrenElements(VariableManager nodeVariableManager, Optional<Node> optionalPreviousNode, NodeDescription nodeDescription,
+            String nodeId, ViewModifier state, ViewModifier parentState, INodeDescriptionRequestor nodeDescriptionRequestor) {
+        List<Element> nodeChildren = new ArrayList<>();
+
+        nodeChildren.addAll(this.getInsideLabel(nodeVariableManager, nodeDescription, nodeId));
+        nodeChildren.addAll(this.getOutsideLabel(nodeVariableManager, nodeDescription, nodeId));
+        nodeChildren.addAll(this.getBorderNodes(optionalPreviousNode, nodeVariableManager, nodeId, state, nodeDescriptionRequestor));
+        nodeChildren.addAll(this.getChildNodes(optionalPreviousNode, nodeVariableManager, nodeId, parentState, nodeDescriptionRequestor));
+
+        return nodeChildren;
+    }
```

Comment:
Technically, this is exactly equivalent to this:

```
const NodeComponent = (props) => {
    const children = nodeChildrenProvider(props);
    return (
        <node>
           {children}
        </node>
    );
}
```

You have a component and instead of having child components, you are using a function which takes props and return elements to compute the children. Could you please provide an API like this instead:

```
const NodeComponent = (props) => {
    return (
       <node>
           <NodeChildrenProvider {...props} />
       </node>
    );
}
```

It would be odd to have a function in React taking props and returning elements but not use it as a component, it is odd here too. And NodeChildrenProvider can still use NodeComponentProps, no issue with that. We would probably not do that in the frontend but it is easier in TypeScript to destructure props and create new props. Given the role of NodeChildrenProvider, handling a part of the concern of NodeComponent, it does make sense to just keep the same props